syft/pkg/cataloger/golang/parse_go_mod.go

```diff
@@ -151,3 +197,445 @@ func parseGoSumFile(resolver file.Resolver, reader file.LocationReadCloser) (map
+	// Example of a module with replace directive: 	k8s.io/kubernetes => k8s.io/kubernetes v1.11.1
+	// {
+	//         "Path": "k8s.io/kubernetes",
+	//         "Version": "v0.17.9",
+	//         "Replace": {
+	//                 "Path": "k8s.io/kubernetes",
+	//                 "Version": "v1.11.1",
+	//                 "Time": "2018-07-17T04:20:29Z",
+	//                 "Dir": "/home/gongyuan_kubeflow_org/go/pkg/mod/k8s.io/kubernetes@v1.11.1",
+	//                 "GoMod": "/home/gongyuan_kubeflow_org/go/pkg/mod/cache/download/k8s.io/kubernetes/@v/v1.11.1.mod"
+	//         },
+	//         "Dir": "/home/gongyuan_kubeflow_org/go/pkg/mod/k8s.io/kubernetes@v1.11.1",
+	//         "GoMod": "/home/gongyuan_kubeflow_org/go/pkg/mod/cache/download/k8s.io/kubernetes/@v/v1.11.1.mod"
+	// }
+	// handle replace directives
+	// Note, we specifically want to replace version field.
+	// Haven't confirmed, but we may also need to override the
+	// entire struct when using replace directive with local folders.
+	tmp := *mod
+	if tmp.Replace != nil {
+		tmp = *tmp.Replace
+	}
+
+	return &tmp
+}
+
+func findLicenseFileLocationsWithFS(dir string, rootDir string, fs afero.Fs) ([]string, error) {
+	dir, err := filepath.Abs(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	rootDir, err = filepath.Abs(rootDir)
+	if err != nil {
+		return nil, err
+	}
+
+	if !strings.HasPrefix(dir, rootDir) {
+		return nil, fmt.Errorf("licenses.Find: rootDir %s should contain dir %s", rootDir, dir)
+	}
+
+	return findAllLicenseCandidatesUpwardsWithFS(dir, licenseRegexp, rootDir, fs)
+}
+
+/*
+findAllLicenseCandidatesUpwards performs a bubble-up search per package because:
+1. pkgInfos represents a sparse vertical distribution of packages within modules
+2. we get more pkgInfos for free when the build configuration is updated
+3. Bubble-up gives us module boundary enforcement and prevents license pollution that could occur on walk down
+
+When we should consider tip to stem:
+- Reduced filesystem calls: Single traversal vs multiple per-package
+- Path deduplication: Avoids re-scanning common parent directories
+- Better for wide module structures: Efficient when many packages share parent paths
+- We need to consider the case here where nested modules are visited by accident and licenses
+are erroneously associated to a 'parent module'; bubble up currently prevents this
+*/
+func findAllLicenseCandidatesUpwardsWithFS(dir string, r *regexp.Regexp, stopAt string, fs afero.Fs) ([]string, error) {
+	// Stop once we go out of the stopAt dir.
+	licenseCandidates := make([]string, 0)
+	visited := make(map[string]bool) // Track visited directories to prevent infinite loops
+
+	for strings.HasPrefix(dir, stopAt) {
+		// Resolve any symlinks to get the actual path
+		// Note: For in-memory filesystems, EvalSymlinks may not work as expected,
+		// but this provides a fallback for real filesystems
+		resolvedDir := dir
+		if _, ok := fs.(*afero.OsFs); ok {
```

Comment:
take a look at https://github.com/spf13/afero/blob/2fb20ad3b2f7a6be83bd142888ea702c3388d3cc/symlink.go#L20-L50 to prevent the need to type assert to the concrete implementation and use the stdlib EvalSymlinks